### PR TITLE
Kafka integration

### PR DESCRIPTION
**Pull request by zapletal-martin**
Purpose of this PR is to add Kafka to deployment and create a Kafka producer to publish events to Kafka for further processing. 

- [X] added basic deployment of Kafka
- [X] added basic deployment of Zookeeper (required by Kafka)
- [X] added first version of code for Kafka producer
- [x] refactor Kafka producer
- [x] refactor Kafka producer actor lifecycle (from current singleton)
- [x] integrate with the rest of the application and remove test endpoint
- [x] Kafka topic lifecycle in Docker container

### CODE REVIEW

File: server/build.sbt [hunk 1]
```diff
@@ -6,8 +6,11 @@ name := "domain"
 //Common code, but not protocols
 lazy val common = project.in(file("common")).dependsOn(contrib)
 
+//Kafka intergration
```
Comment (carlpulley) [inline server/build.sbt]: `integration`

File: server/adapter/src/main/scala/com/eigengo/lift/adapter/AdapterMicroservice.scala [hunk 2]
```diff
@@ -24,7 +25,7 @@ object AdapterBoot {
 
 object AdapterMicroservice extends MicroserviceApp(MicroserviceProps("Adapter")) {
 
-  override def boot(implicit system: ActorSystem, cluster: Cluster): BootedNode = {
+  override def boot(config: Config)(implicit system: ActorSystem, cluster: Cluster): BootedNode = {
```
Comment (carlpulley) [inline server/adapter/src/main/scala/com/eigengo/lift/adapter/AdapterMicroservice.scala]: Whilst I don't have any objections to passing in `config` as a parameter, I do note that it is also possible to extract this value via `system.settings.config`. So, is there a specific reason for having it as an extra argument?

File: server/kafka/src/main/scala/com/eigengo/lift/kafka/KafkaProducerActor.scala [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+  override def kafkaConfig = config.getConfig("kafka.producer")
+
+  private val topic = kafkaConfig.getString("topic")
+
+  private def logs[T](event: T, result: String \/ Unit) =
+    log.info(s"""${self.path} producing "$event" to Kafka with result "$result"""")
+
+  override def receive: Receive = {
+    case event @ _ =>
```
Comment (carlpulley) [inline server/kafka/src/main/scala/com/eigengo/lift/kafka/KafkaProducerActor.scala]: `case event =>`?

File: server/kafka/src/main/scala/com/eigengo/lift/kafka/KafkaProducerActor.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      val produced = produce(event, topic)
+      logs(event, produced)
+  }
+
+  override def postStop() = {
+    close
+    postStop
```
Comment (carlpulley) [inline server/kafka/src/main/scala/com/eigengo/lift/kafka/KafkaProducerActor.scala]: Should this be `super.postStop`?

File: server/adapter/src/main/scala/com/eigengo/lift/adapter/AdapterMicroservice.scala (hunk 2 above)
Comment (zapletal-martin) [inline server/adapter/src/main/scala/com/eigengo/lift/adapter/AdapterMicroservice.scala]: Passing `config` to a microservice's `boot()` method as a parameter feels more general. In our case where the config is shared between akka (or the actorsystem) and the Kafka producer I agree it does not matter. It however feels a bit hidden and I thought passing the `config` explicitly for the microservice to have it visibly available without relying on the system would be nicer.

If you think using `system.settings.config` would be a better idea I can change that.

File: server/kafka/src/main/scala/com/eigengo/lift/kafka/KafkaProducerActor.scala (hunk 4 above)
Comment (zapletal-martin) [inline server/kafka/src/main/scala/com/eigengo/lift/kafka/KafkaProducerActor.scala]: Ah thanks calling postStop could be nasty!

File: server/adapter/src/main/scala/com/eigengo/lift/adapter/AdapterMicroservice.scala (hunk 2 above)
Comment (carlpulley) [inline server/adapter/src/main/scala/com/eigengo/lift/adapter/AdapterMicroservice.scala]: Thanks for that. I was more interested in your rationale here.

**Comment by janm399** [comment]
Does this make any change to how we run the server locally—that is does ``sbt "project main" run`` still work? Or do I need to do anything else to get it running? Also, have you tried it with the live iOS application submitting real data?

**Comment by zapletal-martin** [comment]
No it does not make any change to how we run the server locally. The kafka client is set to async so if you don't have local Kafka broker it will just drop the messages.

The integration with the app is currently only publishing some events. It however provides methods to publish any events we want. We first need to figure out what we want in Kafka.

I have not tried with the iOS app. Just by hitting the rest api using postman (which should be the same).

**Pull request merged**